### PR TITLE
Fix ticket pricing constraint

### PR DESCRIPTION
**Pull request by Timo614**
Fixes: #414

This issue fixes the associated issue by fixing some faulty logic in the SQL constraint that was being used to prevent overlapping periods. The logic was checking if there was a period that overlapped the end date passed in but since end dates can equal the value for start dates (to make them continuous) it was incorrectly seeing an overlap when one wasn't there. I updated the tests to catch the issue that triggered this as well.

In addition this branch also fixes:
1. Error related application helpers now properly set an error on the http response (triggering a rollback on the transaction)
2. Updated validation logic for several controllers to use the conversion logic we have in place for db errors (pushing their validations into their models's saving logic and allowing the controllers to automatically return the validation error from the result in a standardized way).
3. Removes the constraint used here in favor of a validation that occurs after all of the ticket pricing rows have been updated. I noticed while fixing this function that since we update all of these at once they can conflict if we're moving a period around and they intersect prior to the others updating. We may have been able to use a trigger and a call with multiple updates but it would make the logic pretty messy as we're removing, adding, and updating in the same logic and diesel doesn't support updating multiple without using custom SQL. This approach rolls the transaction back if it runs into an issue so it accomplishes the same while still allowing us to have readable code for this. Could always refactor it later though if we have a preference for the other approach (which I think would work but I didn't test it).
4. Throws an error if an invalid ticket pricing ID is passed in for updating (previously it just swallowed the error so confused me a bit when debugging the issue and I used the wrong ID)

The new logic to check for overlaps is the following (I think this should cover it now but would be good for anyone reviewing to once over it to make sure it's not flawed -- the start date is inclusive but the end date isn't):

```
            SELECT id
            FROM ticket_pricing
            WHERE
                -- Filter out current record being updated
                ID <> $1
            AND
                -- Filter to the current ticket type
                ticket_type_id = $2
            AND
            (
                -- Does any period overlap the start date
                (start_date <= $3 AND end_date > $3)
            OR
                -- Does any period overlap the end date
                (start_date < $4 AND end_date >= $4)
            OR
                -- Does this period completely overlap another period
                (start_date >= $3 AND end_date <= $4)
            )
```

Examples:

Adding ticket pricing:

```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiJlMTdmMTczMC02NmI0LTQzN2EtOGQwNS1hODEzNWQzYmI1YWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjE0NTQ2fQ./5CWfEQkriCDhWp7kpd89nm0O4sxqQ/VV2Gr757iirY" http://localhost:9000/events/840d32f4-3c77-4183-b81f-bd298fec399d/ticket_types -X POST --data '{"event_id":"840d32f4-3c77-4183-b81f-bd298fec399d","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"name":"Early worms","price_in_cents":1000,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-16T10:00:00.000"},{"name":"Late birds","price_in_cents":1200,"start_date":"2018-10-16T10:00:00.000","end_date":"2018-10-20T10:00:00.000"}]}'
{"id":"5ae1cc48-459e-4fde-8d82-4a57caa22626"}
```

Updating ticket pricing:

```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiJlMTdmMTczMC02NmI0LTQzN2EtOGQwNS1hODEzNWQzYmI1YWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjE0NTQ2fQ./5CWfEQkriCDhWp7kpd89nm0O4sxqQ/VV2Gr757iirY" http://localhost:9000/events/840d32f4-3c77-4183-b81f-bd298fec399d/ticket_types/5ae1cc48-459e-4fde-8d82-4a57caa22626 -X PATCH --data '{"event_id":"840d32f4-3c77-4183-b81f-bd298fec399d","id":"5ae1cc48-459e-4fde-8d82-4a57caa22626","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"17ed7392-3c57-41ce-bcc4-eddf5a1b7b26","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-16T10:00:00.000"},{"id":"9fb4fbed-08bb-4a41-a3a0-7b499175071d","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-16T10:00:00.000","end_date":"2018-10-21T10:00:00.000"}]}' -i
HTTP/1.1 200 OK
```

Swapping ticket pricing (previously didn't work given the logic updates one ticket pricing at a time and constraints are single row bound):

```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI4MzBiNzZiMC1jNjlmLTQ5YWEtODE1My1iNjNlYTI2OWZmNzIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjM3MDQ2fQ.Axf9iwhTcTOL4AxXw8+5UoS9gqR4X5NcQXCsXSs/0IM" http://localhost:9000/events/b10dad68-8b18-4af8-83c2-dec184120a01/ticket_types/4c47228c-400e-4c2b-a6e5-824d292ec0c8 -X PATCH --data '{"event_id":"b10dad68-8b18-4af8-83c2-dec184120a01","id":"4c47228c-400e-4c2b-a6e5-824d292ec0c8","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"63b7a4cb-4fff-4eb9-8e57-c9483043672a","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-17T10:00:00.000","end_date":"2018-10-21T10:00:00.000"},{"id":"2ec258ac-930f-4fdd-90aa-4d938955d6e3","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-17T10:00:00.000"}]}' -i

select id, start_date, end_date from ticket_pricing where ticket_type_id = '4c47228c-400e-4c2b-a6e5-824d292ec0c8';
                  id                  |     start_date      |      end_date       
--------------------------------------+---------------------+---------------------
 63b7a4cb-4fff-4eb9-8e57-c9483043672a | 2018-10-17 10:00:00 | 2018-10-21 10:00:00
 2ec258ac-930f-4fdd-90aa-4d938955d6e3 | 2018-10-15 10:00:00 | 2018-10-17 10:00:00

curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI4MzBiNzZiMC1jNjlmLTQ5YWEtODE1My1iNjNlYTI2OWZmNzIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjM3MDQ2fQ.Axf9iwhTcTOL4AxXw8+5UoS9gqR4X5NcQXCsXSs/0IM" http://localhost:9000/events/b10dad68-8b18-4af8-83c2-dec184120a01/ticket_types/4c47228c-400e-4c2b-a6e5-824d292ec0c8 -X PATCH --data '{"event_id":"b10dad68-8b18-4af8-83c2-dec184120a01","id":"4c47228c-400e-4c2b-a6e5-824d292ec0c8","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"63b7a4cb-4fff-4eb9-8e57-c9483043672a","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-17T10:00:00.000"},{"id":"2ec258ac-930f-4fdd-90aa-4d938955d6e3","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-17T10:00:00.000","end_date":"2018-10-21T10:00:00.000"}]}' -i
HTTP/1.1 200 OK

select id, start_date, end_date from ticket_pricing where ticket_type_id = '4c47228c-400e-4c2b-a6e5-824d292ec0c8';
                  id                  |     start_date      |      end_date       
--------------------------------------+---------------------+---------------------
 63b7a4cb-4fff-4eb9-8e57-c9483043672a | 2018-10-15 10:00:00 | 2018-10-17 10:00:00
 2ec258ac-930f-4fdd-90aa-4d938955d6e3 | 2018-10-17 10:00:00 | 2018-10-21 10:00:00
```

Behavior when overlap occurs (reverts transaction):

```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI4MzBiNzZiMC1jNjlmLTQ5YWEtODE1My1iNjNlYTI2OWZmNzIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjM3MDQ2fQ.Axf9iwhTcTOL4AxXw8+5UoS9gqR4X5NcQXCsXSs/0IM" http://localhost:9000/events/b10dad68-8b18-4af8-83c2-dec184120a01/ticket_types/4c47228c-400e-4c2b-a6e5-824d292ec0c8 -X PATCH --data '{"event_id":"b10dad68-8b18-4af8-83c2-dec184120a01","id":"4c47228c-400e-4c2b-a6e5-824d292ec0c8","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"63b7a4cb-4fff-4eb9-8e57-c9483043672a","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-21T10:00:00.000"},{"id":"2ec258ac-930f-4fdd-90aa-4d938955d6e3","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-17T10:00:00.000"}]}' -i
HTTP/1.1 422 Unprocessable Entity

{"error":"Validation error","fields":{"ticket_pricing":[{"code":"ticket_pricing_overlapping_periods","message":null,"params":{"end_date":"2018-10-21T10:00:00","id":"63b7a4cb-4fff-4eb9-8e57-c9483043672a","start_date":"2018-10-15T10:00:00","ticket_type_id":"4c47228c-400e-4c2b-a6e5-824d292ec0c8"}},{"code":"ticket_pricing_overlapping_periods","message":null,"params":{"end_date":"2018-10-17T10:00:00","id":"2ec258ac-930f-4fdd-90aa-4d938955d6e3","start_date":"2018-10-15T10:00:00","ticket_type_id":"4c47228c-400e-4c2b-a6e5-824d292ec0c8"}}]}}

select id, start_date, end_date from ticket_pricing where ticket_type_id = '4c47228c-400e-4c2b-a6e5-824d292ec0c8';
                  id                  |     start_date      |      end_date       
--------------------------------------+---------------------+---------------------
 63b7a4cb-4fff-4eb9-8e57-c9483043672a | 2018-10-15 10:00:00 | 2018-10-17 10:00:00
 2ec258ac-930f-4fdd-90aa-4d938955d6e3 | 2018-10-17 10:00:00 | 2018-10-21 10:00:00
```

Previous message (no longer applicable but I mentioned it in a comment below so persisting it):

> Thinking things over though I wonder if we should just entirely remove the constraint. We allow updating of ticket types as a whole unit so multiple can be sent with new start and end dates. The logic in the db though updates these one at a time so it can lead to issues with the constraint depending upon the order updated. So if the first one tries to extend its end date past the start date of the second one and the second moves its start date further on where it wouldn't conflict it can still fail as the first now intersects the second's existing record until its been updated.

> We have two options:
> 1. Remove the constraint without replacement (any db validation will also see this problem as it's an ordering issue)
> 2. Update these records in tandem somehow on the diesel side instead of one by one (not sure how to accomplish that but from looking at stackoverflow it may be doable)

> This can be seen here:

> Fails attempting to update both at once:
```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiJlMTdmMTczMC02NmI0LTQzN2EtOGQwNS1hODEzNWQzYmI1YWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjE0NTQ2fQ./5CWfEQkriCDhWp7kpd89nm0O4sxqQ/VV2Gr757iirY" http://localhost:9000/events/840d32f4-3c77-4183-b81f-bd298fec399d/ticket_types/5ae1cc48-459e-4fde-8d82-4a57caa22626 -X PATCH --data '{"event_id":"840d32f4-3c77-4183-b81f-bd298fec399d","id":"5ae1cc48-459e-4fde-8d82-4a57caa22626","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"17ed7392-3c57-41ce-bcc4-eddf5a1b7b26","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-17T10:00:00.000"},{"id":"9fb4fbed-08bb-4a41-a3a0-7b499175071d","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-17T10:00:00.000","end_date":"2018-10-21T10:00:00.000"}]}' -i
HTTP/1.1 500 Internal Server Error
```

> Update second date first alone:
```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiJlMTdmMTczMC02NmI0LTQzN2EtOGQwNS1hODEzNWQzYmI1YWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjE0NTQ2fQ./5CWfEQkriCDhWp7kpd89nm0O4sxqQ/VV2Gr757iirY" http://localhost:9000/events/840d32f4-3c77-4183-b81f-bd298fec399d/ticket_types/5ae1cc48-459e-4fde-8d82-4a57caa22626 -X PATCH --data '{"event_id":"840d32f4-3c77-4183-b81f-bd298fec399d","id":"5ae1cc48-459e-4fde-8d82-4a57caa22626","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"17ed7392-3c57-41ce-bcc4-eddf5a1b7b26","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-16T10:00:00.000"},{"id":"9fb4fbed-08bb-4a41-a3a0-7b499175071d","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-17T10:00:00.000","end_date":"2018-10-21T10:00:00.000"}]}' -i
HTTP/1.1 200 OK
```

> Update first date now logic matches the first failing one:
```
curl --header "Content-Type: application/json" --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiJlMTdmMTczMC02NmI0LTQzN2EtOGQwNS1hODEzNWQzYmI1YWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM5NjE0NTQ2fQ./5CWfEQkriCDhWp7kpd89nm0O4sxqQ/VV2Gr757iirY" http://localhost:9000/events/840d32f4-3c77-4183-b81f-bd298fec399d/ticket_types/5ae1cc48-459e-4fde-8d82-4a57caa22626 -X PATCH --data '{"event_id":"840d32f4-3c77-4183-b81f-bd298fec399d","id":"5ae1cc48-459e-4fde-8d82-4a57caa22626","name":"GA","capacity":100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-20T10:00:00.000","ticket_pricing":[{"id":"17ed7392-3c57-41ce-bcc4-eddf5a1b7b26","name":"Early worms","price_in_cents":1100,"start_date":"2018-10-15T10:00:00.000","end_date":"2018-10-17T10:00:00.000"},{"id":"9fb4fbed-08bb-4a41-a3a0-7b499175071d","name":"Late birds","price_in_cents":1400,"start_date":"2018-10-17T10:00:00.000","end_date":"2018-10-21T10:00:00.000"}]}' -i
HTTP/1.1 200 OK
```

### CODE REVIEW

**Comment by Timo614** [comment]
Going to see how hard it would be to get multiple updates here working to avoid that situation I mentioned above in the description where if you update multiple record periods to new ones they can hit the constraint depending upon the order the updates are executed in. I haven't used that syntax before and not sure if diesel supports it so may have to be something more custom.

**Comment by Timo614** [comment]
Yeah multiple updates do not get around the issue as it appears the constraint is checked per record as updates happen. It looks like going the constraint route isn't the best option here. One thing I did think of though is since we have the transactions revert on error we could just run the validation after the records have both been updated (as some new function) and then trigger an error condition then which would force the transaction to revert and look the same to the frontend. I'll go that approach as I think it should work even in this case as it'll happen after both have been adjusted.

**Comment by Timo614** [comment]
Just a heads up @Jasonvdb the validation responses were standardized to all use `UNPROCESSABLE_ENTITY` (code: 422) in this branch. The format they used to use is still the same just the different code as we had some using `BAD_REQUEST` (code: 400) for some of the others.

File: api/src/helpers/application.rs [hunk 1]
```diff
@@ -1,39 +1,53 @@
...
-use validator::ValidationErrors;
 
 pub fn unauthorized() -> Result<HttpResponse, BigNeonError> {
     unauthorized_with_message("Unauthorized")
 }
 
 pub fn unauthorized_with_message(message: &str) -> Result<HttpResponse, BigNeonError> {
     warn!("Unauthorized: {}", message);
-    Ok(HttpResponse::Unauthorized().json(json!({"error": message.to_string()})))
+    let error: BigNeonError = AuthError {
```
Comment (mikethetike) [inline api/src/helpers/application.rs]: I see from your PR comments why you did this, but a new dev might see this code and reduce it. Maybe add a comment here to state that it is necessary to rollback the transaction

**Pull request merged**